test/cuda_sims.jl

```diff
@@ -0,0 +1,344 @@
+using Catlab
+using Decapodes
+using DiagrammaticEquations
+using CombinatorialSpaces
+using GeometryBasics
+using MLStyle
+using ComponentArrays
+using OrdinaryDiffEq
+using CUDA
+using CUDA.CUSPARSE
+using LinearAlgebra
+using SparseArrays
+using Statistics
+using Distributions
+using Test
+Point2D = Point2{Float64}
+Point3D = Point3{Float64}
+CUDA.allowscalar(false)
+
+function RMSE(a, b)
+  return sqrt(mean((a .- b).^2))
+end
+
+@testset "Heat Equation Float64" begin
+  Heat = @decapode begin
+    U::Form0
+    ∂ₜ(U) == 100 * Δ(U)
+  end
+
+  tₑ = 8.5
+  constants_and_parameters = ()
+
+  s = loadmesh(Icosphere(4))
+  sd = EmbeddedDeltaDualComplex2D{Bool, Float64, Point3D}(s)
+  subdivide_duals!(sd, Circumcenter())
+
+  U = map(sd[:point]) do (x,_)
+    return x
+  end
+
+  # CPU Setup and Solve
+  cpu_sim = eval(gensim(Heat))
+  cpu_fₘ = cpu_sim(sd, nothing, DiagonalHodge())
+
+  cpu_u₀ = ComponentArray(U=U)
+
+  prob = ODEProblem(cpu_fₘ, cpu_u₀, (0, tₑ), constants_and_parameters)
+  cpu_soln = solve(prob, Tsit5(), save_everystep=false)
+
+  # CUDA Setup and Solve
+  cuda_sim = eval(gensim(Heat, code_target=cuda()))
+  cuda_fₘ = cuda_sim(sd, nothing, DiagonalHodge())
+  
+  cuda_u₀ = ComponentArray(U=CuArray{Float64}(U))
+  prob = ODEProblem(cuda_fₘ, cuda_u₀, (0, tₑ), constants_and_parameters)
+  cuda_soln = solve(prob, Tsit5(), save_everystep=false)
+
+  @test all(isapprox(cpu_soln(tₑ).U, Array(cuda_soln(tₑ).U); atol=1e-12))
+  @test RMSE(cpu_soln(tₑ).U, Array(cuda_soln(tₑ).U)) < 1e-13 
+end
+
+@testset "Heat Equation Float32" begin
+  Heat = @decapode begin
+    U::Form0
+    ∂ₜ(U) == 100 * Δ(U)
+  end
+
+  tₑ = 11.5f0
+  constants_and_parameters = ()
+
+  s = loadmesh(Icosphere(3))
+  sd = EmbeddedDeltaDualComplex2D{Bool, Float32, Point3{Float32}}(s)
+  subdivide_duals!(sd, Circumcenter())
+
+  U = map(sd[:point]) do (x,_)
+    return x
+  end
+
+  # CPU Setup and Solve
+  cpu_sim = eval(gensim(Heat, stateeltype=Float32))
+  cpu_fₘ = cpu_sim(sd, nothing, DiagonalHodge())
+
+  cpu_u₀ = ComponentArray(U=U)
+
+  prob = ODEProblem(cpu_fₘ, cpu_u₀, (0, tₑ), constants_and_parameters)
+  cpu_soln = solve(prob, Tsit5(), save_everystep=false)
+
+  # CUDA Setup and Solve
+  cuda_sim = eval(gensim(Heat, code_target=cuda(), stateeltype=Float32))
+  cuda_fₘ = cuda_sim(sd, nothing, DiagonalHodge())
+  
+  cuda_u₀ = ComponentArray(U=CuArray{Float32}(U))
+  prob = ODEProblem(cuda_fₘ, cuda_u₀, (0, tₑ), constants_and_parameters)
+  cuda_soln = solve(prob, Tsit5(), save_everystep=false)
+
+  @test all(isapprox(cpu_soln(tₑ).U, Array(cuda_soln(tₑ).U); atol=1e-5))
+  @test RMSE(cpu_soln(tₑ).U, Array(cuda_soln(tₑ).U)) < 1e-6
+end
+
+@testset "Brusselator Float64" begin
+  Brusselator = @decapode begin
```

Comment:
Canon

Reply:
Turn this into an issue and another PR since there are tests in other files that should also use canon.